### DISCUSSION TITLE
Questions about some logging at the warning level

### DISCUSSION THREAD

**Discussion by facewindu** [oshi/oshi#2721] [General]
We are currently seeing [this logging](https://github.com/oshi/oshi/blob/0ae5c65613e7b41752b83b7bfec0c24914482c2a/oshi-core/src/main/java/oshi/util/platform/windows/PerfCounterQuery.java#L185-L188) and [this logging](https://github.com/oshi/oshi/blob/0ae5c65613e7b41752b83b7bfec0c24914482c2a/oshi-core/src/main/java/oshi/util/platform/windows/PerfCounterQuery.java#L78).

```

Unable to localize Process performance counter.  Error 0xc0000bbd.
Failed to locate performance object for Process in the registry. Performance counters may be corrupt. Pdh call failed with error code 0xC0000BB8 
Disabling further attempts to query Process. 
Unable to localize Processor Information performance counter.  Error 0xc0000bbd. 
Failed to locate performance object for Processor Information in the registry. Performance counters may be corrupt. Pdh call failed with error code 0xC0000BB8 
Disabling further attempts to query Processor Information. 
Unable to localize PhysicalDisk performance counter.  Error 0xc0000bbd. 
Failed to locate performance object for PhysicalDisk in the registry. Performance counters may be corrupt. Pdh call failed with error code 0xC0000BB8 
Disabling further attempts to query PhysicalDisk.
```
It seems to me that
- the former is about creating a localized string. Is it really worth logging this at the warning level (debug would be fine ?)
- the latter has a fallback strategy using WMI. So same question, to me it would be better logged as debug

**Comment by dbwiddis** [comment]
> Is it really worth logging this at the warning level (debug would be fine ?)

At least one of these three messages should be at the warning level.  The target operating system has corrupt counters that have triggered an exception handling path and are impacting, at a minimum, OSHI's performance requiring it to use a fallback.  I don't want to hide the answer to "why is my system slower" answers in debug.

> * the latter has a fallback strategy using WMI. So same question, to me it would be better logged as debug

This is really just the reaction of the system to the preceding corrupt counter message.  Perhaps the log entries could be appended to a single string?  It's probably worth at least INFO logging if not WARN.

Feel free to submit a PR to lower the "disabling" message to INFO, I'll approve it.  But all of these are one-time log messages, usually closer to startup.

There are configuration options in the config file to remove some logging when people knowingly disable performance counters, perhaps a similar configuration could apply here, but this doesn't seem like the same type of "configuration by choice".  The system is in an abnormal state and I think users should be warned about that.

**Reply by facewindu** [reply]
I guess my main concern is that if the consumer is periodically calling Oshi, and the system is in such a bad state, the log will contain many duplicated entries like that.
Maybe this means that Oshi would benefit from having some logic to log such things only once, or expose some health information so that consumers can check and call things that are in a bad state ony once.

**Reply by dbwiddis** [reply]
> I guess my main concern is that if the consumer is periodically calling Oshi, and the system is in such a bad state, the log will contain many duplicated entries like that. Maybe this means that Oshi would benefit from having some logic to log such things only once, or expose some health information so that consumers can check and call things that are in a bad state ony once.

All of the above log messages originate in a path that will only execute once per counter.

Here's the wrapper method inside which all 3 log warnings above are generated:
https://github.com/oshi/oshi/blob/ede581c145dec3a6fe2ee9d807c0f9df354d523e/oshi-core/src/main/java/oshi/util/platform/windows/PerfCounterWildcardQuery.java#L80-L95

Note that we check whether the query has failed before, before attempting any more PDH Counter retrievals for that counter.
https://github.com/oshi/oshi/blob/ede581c145dec3a6fe2ee9d807c0f9df354d523e/oshi-core/src/main/java/oshi/util/platform/windows/PerfCounterWildcardQuery.java#L41

That is a static variable, so that set will remain until the entire JVM is restarted.  `queryInstancesAndValuesFromPDH()` takes us eventually here:
https://github.com/oshi/oshi/blob/ede581c145dec3a6fe2ee9d807c0f9df354d523e/oshi-core/src/main/java/oshi/util/platform/windows/PerfCounterWildcardQuery.java#L144-L150

And the first error is generated on the PDH failure call.

I believe logging a warning on an unexpected system error is appropriate.   I recognize there is duplication in the first and second log messages, but they only happen once per counter name per JVM instance, and the second message explains the meaning of the first and provides detailed information (corrupt performance counters) for which there is ample information available how to resolve.  

The last message is, perhaps, informational and could be reduced from a warning to INFO or even DEBUG, but the three messages together do somewhat tell a story. OSHI is saying "your registry is corrupt, and I'm handling it with a backup".

Now in theory you might say "hey if one set of counters is corrupt then all of them are" and that is probalby the most common scneario.  We could certainly add some logic (or a configuration vairable) to say "log at most N PDH failures".

**Reply by dbwiddis** [reply]
As the above reply is a bit wordy, I'll tack on the TLDR:
 - Currently there are 3 WARN log messages for each counter
 - I still think we need at least 1 WARN log message and would favor the second one (describing corrupt counters). As the first one is easily disabled with logging filters, I'm not overly eager to disable it; however given that it almost always results in the second message I could be convinced to lower it to debug.
 - I'm ok with lowering the third message to INFO.

I'm also ok with adding a configuration setting to assume if any counter hits these issues, all should be disabled.  The net effect would be a single log message similar to the 2nd ("... corrupt ....") but without naming the specific counter.   That combined with the previous suggestions would reduce the 12 messages above to one single WARN message, if appropriately configured.

**Reply by facewindu** [reply]
I'm fine with your last message. We hit the "if one set of counters is corrupt then all of them are" and saw multiple messages with slightly different counter names.
Thanks for the detailed answer.

**Reply by dbwiddis** [reply]
Great.  I'll try to get something together by this weekend and cut a release as there's another pending logging change.

**Reply by dbwiddis** [reply]
6.6.4 release cut, see #2725
